### PR TITLE
[FLINK-36379] Improve (Global)Committer with UC disabled [1.20]

### PR DESCRIPTION
**Pull request by AHeise**
Backport of #25456 with some adjustments in `GlobalCommitterOperater` (main doesn't have `GlobalCommitter` code anymore).

### CODE REVIEW

**Comment by davidradl** [comment]
Reviewed by Chi on 21/11/24. Looks in hand, test failures currently

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        transformation.setMaxParallelism(1);
+        copySafely(transformation::setName, globalCommitterTransform::getName);
+        copySafely(transformation::setUid, globalCommitterTransform::getUid);
+        copySafely(transformation::setUidHash, globalCommitterTransform::getUserProvidedNodeHash);
+
+        return Arrays.asList(global.getId(), transformation.getId());
+    }
+
+    private static <T> void copySafely(Consumer<T> consumer, Supplier<T> provider) {
+        T value = provider.get();
```
Comment (davidradl) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java:102]: I am wondering 
- are name, uid or uidHash nullable? The uid seems like a mandatory non nullable identifier. 
- should we error if any of these values are null? 
- or is there a valid case for having any of these values as null, if so should we not copy over the null?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        return translateInternal(transformation, true);
+    }
+
+    @Override
+    public Collection<Integer> translateForStreaming(
+            GlobalCommitterTransform<CommT> transformation, Context context) {
+        return translateInternal(transformation, false);
+    }
+
+    private Collection<Integer> translateInternal(
```
Comment (davidradl) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java:68]: should we add junits for this method - or do you think the higher level tests already cover us?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java [hunk 3]
```diff
@@ -181,7 +181,7 @@ private void commitAndEmitCheckpoints() throws IOException, InterruptedException
     private void commitAndEmit(CheckpointCommittableManager<CommT> committableManager)
             throws IOException, InterruptedException {
         Collection<CommittableWithLineage<CommT>> committed = committableManager.commit(committer);
```
Comment (davidradl) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:183]: NIT: as we are no longer using committed in the if, we could move `committableManager.commit(committer)` inlined to line 189 and remove the `committed` variable

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManager.java [hunk 4]
```diff
@@ -56,14 +56,20 @@ public interface CheckpointCommittableManager<CommT> {
...
+     */
+    boolean hasGloballyReceivedAll();
+
     /**
      * Commits all due committables if all respective committables of the specific subtask and
      * checkpoint have been received.
      *
      * @param committer used to commit to the external system
      * @return successfully committed committables with meta information
-     * @throws IOException
```
Comment (davidradl) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManager.java:65]: why are these removed, we still throw these exceptions.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+        Set<Integer> seenIds = new HashSet<>();
+        Queue<Transformation<?>> pendingsTransformations =
+                new ArrayDeque<>(Collections.singleton(dsTransformation));
+        while (!pendingsTransformations.isEmpty()) {
+            Transformation<?> transformation = pendingsTransformations.poll();
+            if (transformation instanceof OneInputTransformation) {
+                StreamOperatorFactory<?> operatorFactory =
+                        ((OneInputTransformation<?, ?>) transformation).getOperatorFactory();
+                if (operatorFactory instanceof CommitterOperatorFactory) {
```
Comment (davidradl) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java:123]: I think a comment to describe why we return true here would be useful.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java [hunk 6]
```diff
@@ -188,10 +195,19 @@ private void expand() {
             }
         }
 
+        private <R> void repeatUntilConverged(Supplier<R> producer) {
+            R lastResult = producer.get();
+            R nextResult;
+            while (!lastResult.equals(nextResult = producer.get())) {
```
Comment (davidradl) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java:201]: can we get nulls here?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java:102]: Yes all of the three are nullable and that's the default value. Some of the setters throw error on null, hence we can't copy the null. So this is the safest way to "copy" the value.

Any assertions on the presence of the values is also a bit more difficult as they are set at different point during the transformation and the GlobalCommitterTransformation is added usually as part of the sink transformation, which is then recursively executed. It's a bit convoluted and only necessary because the transformation is the only place for us to figure out if we are running in streaming or batch mode...

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java (hunk 2 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java:68]: There are a couple of low level ITs (Sink(V1/V)TransformationTranslatorITCase) tests that assert like unit tests (is name/uid/uidHash correctly copied) and plenty of high-level ITs (testing batch/checkpointing settings). So I don't see the immediate use to augment it with junit. Coverage is very high.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java (hunk 3 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:183]: `commit` is unfortunately not side-effect free; we would never proceed with your proposal. But maybe it should be side-effect free. 

However, I'd like to pick that idea up on master and not the backport (this PR).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManager.java (hunk 4 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManager.java:65]: This gives us Spotless warnings. The general consent is to either document when exceptions are thrown or not document it. Since checked exceptions are part of the signature, the javadoc still shows them as being thrown. So leaving the tags incomplete is just redundant.

Note for public API we aim for documenting all checked exceptions.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java [hunk 7]
```diff
@@ -160,13 +169,37 @@ public Collection<CommittableWithLineage<CommT>> commit(Committer<CommT> committ
         return committed;
     }
 
-    Collection<CommitRequestImpl<CommT>> getPendingRequests(boolean onlyIfFullyReceived) {
+    Collection<CommitRequestImpl<CommT>> getPendingRequests(boolean assertFull) {
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java:172]: That's a good idea. However, the next (and last) PR is touching the code again and removes the flag. It's only here temporarily to cut the PRs smaller. The next PR removes async retries (which should have never be done) and that simplifies the invariants significantly.

**Review by fapaul — APPROVED** [review]
I already reviewed the changes on master and verified that semantically they are the same.
